utils/nce_neighbors.py

```diff
@@ -0,0 +1,46 @@
+import random
+import numpy as np
+import heapq
+import operator
+
+import torch
+from torchtext import data
+
+# get the nearest negative samples to the positive sample by computing the feature difference
+def get_nearest_neg_id(pos_feature, neg_dict, distance="cosine", k=1):
+    dis_list = []
+    pos_feature = pos_feature.data.cpu().numpy()
+    pos_feature_norm = pos_feature / np.sqrt(sum(pos_feature ** 2))
+    neg_list = []
+    for key in neg_dict:
+        if distance == "l2":
+            dis = np.sqrt(np.sum((np.array(pos_feature) - neg_dict[key]["feature"]) ** 2))
+        elif distance == "cosine":
+            neg_feature = np.array(neg_dict[key]["feature"])
+            feat_norm = neg_feature / np.sqrt(sum(neg_feature ** 2))
+            dis = 1 - feat_norm.dot(pos_feature_norm)
+        dis_list.append(dis)
+        neg_list.append(key)
+
+    k = min(k, len(neg_dict))
+    min_list = heapq.nsmallest(k, enumerate(dis_list), key=operator.itemgetter(1))
+    min_id_list = [neg_list[x[0]] for x in min_list]
+    return min_id_list
+
+# get the negative samples randomly
+def get_random_neg_id(q2neg, qid_i, k=5):
+    # question 1734 has no neg answer
```

Comment:
which dataset is the comment referring to?

Reply:
TrecQA. I will comment it.